### PR TITLE
sensor: eeprom: Add TMP116 EEPROM access

### PR DESCRIPTION
**Pull request by kurddt**
Read and write accesses are supported by using
either custom or eeprom API

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
> Only read-only access is supported by using either custom or eeprom API

Thank you. How come only read-only access is implemented?

**Comment by kurddt** [comment]
Mainly because it's the only thing I used and that I tested. It allows to read the IDs preprogrammed in the sensor eeprom.

If I'm able clear some time I will add the write part, but I hope this will not block the PR

File: drivers/sensor/tmp116/tmp116.c
Comment (kurddt) [inline drivers/sensor/tmp116/tmp116.c:66]: I'm not sure about this part. Is it better to do as if the TMP117 have a contiguous EEPROM of 3*16bits instead of 4*16bits with reserved 16bits

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/sensor/tmp116/Kconfig
Comment (kurddt) [reply drivers/sensor/tmp116/Kconfig]: For my own understanding what's the difference between the two construction and when it is advised to use one or the other ?

Comment (henrikbrixandersen) [reply drivers/sensor/tmp116/Kconfig]: The recommendation is to use `depends on` if it's just a single option, but an `if` block if multiple options depend on the same option.

File: include/drivers/sensor/tmp116.h
Comment (kurddt) [reply include/drivers/sensor/tmp116.h:11]: Only for `off_t` I can change it to `sys/types.h` if you prefer

Comment (henrikbrixandersen) [reply include/drivers/sensor/tmp116.h:11]: Please do.

**Review by henrikbrixandersen — APPROVED** [review]
Looks good, thank you. One minor nit, not a blocker.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
@MaureenHelm Are you ok with the changes to the sensor driver?

File: drivers/sensor/tmp116/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/tmp116/Kconfig:12]: Is this really needed? I've been working on reducing the number of driver-specific init priority configs and instead using common configs for each driver class (e.g., SENSOR_INIT_PRIORITY, EEPROM_INIT_PRIORITY, etc.)

Comment (kurddt) [reply drivers/sensor/tmp116/Kconfig:12]: out of curiosity what is the reason for limiting those driver-specific priorities ? 
Usually for my applications I end up with a lot of driver-specific init priority defined otherwise because of devices dependency some initialization fails. That said it's probably less true for "leaf" devices like sensors or eeprom. I will review the usage.

In any case I will change the default value of `EEPROM_TMP116_INIT_PRIORITY` to `EEPROM_INIT_PRIORITY`

Comment (kurddt) [reply drivers/sensor/tmp116/Kconfig:12]: I think this config is not really required in my application so I will remove it as per your suggestion. `EEPROM_INIT_PRIORITY` doesn't seem to exist so it's propably why I set `EEPROM_TMP116_INIT_PRIORITY` default to an arbitrary value

Comment (MaureenHelm) [reply drivers/sensor/tmp116/Kconfig]: Please see #40044 which adds `EEPROM_INIT_PRIO`.

This cleanup of initialization priorities was triggered by #38631. Many drivers use `KERNEL_INIT_PRIORITY_DEVICE` directly, which can cause problems if you need to initialize two drivers in a particular order, in that case it was a clock driver and a UART driver. We needed a little more granularity to specify initialization priorities, but didn't want to introduce the Kconfig symbol sprawl that would come with specifying priorities for every single driver. It will also make applications more portable across multiple boards.

Comment (kurddt) [reply drivers/sensor/tmp116/Kconfig]: @MaureenHelm I rebased on main after  #40044 was merged and removed the driver specific init priority in favor of `EEPROM_INIT_PRIORITY`

**Comment by kurddt** [comment]
@MaureenHelm Could you review again ?

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
